### PR TITLE
Fix split_state_dict function for MoE models

### PR DESCRIPTION
**Pull request by eternally-z**
# What does this PR do ?

main PR: https://github.com/NVIDIA/Megatron-LM/pull/3676

This PR fixes an issue in _split_state_dict, which splits the state dict into sub-state dicts for chained optimizers.

The previous offset-based implementation failed in MoE scenarios. Since MoE parts often have a separate optimizer, the offset-based splitting logic caused chunk mis-assignment. As a result, the MoE optimizer was unable to obtain the correct state_dicts. 

@AniZpZ

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/optimizer/optimizer.py [hunk 1]
```diff
@@ -1161,20 +1161,25 @@ def _split_state_dict(self, state_dict):
         state_dicts = [None] * len(self.chained_optimizers)
         if state_dict is not None:
             if len(self.model_chunks) == 1:
-                state_dicts[0] = state_dict
+                # When there is only one global model chunk, all sub-optimizers
+                # (e.g., dense and MoE parts) use the same model state dict.
+                state_dicts = [state_dict] * len(self.chained_optimizers)
```
Comment (kunlunl) [inline megatron/core/optimizer/optimizer.py:1166]: This `* len(xxx)` doesn't calls deep copy. Will it trigger any bug?

File: megatron/core/optimizer/optimizer.py [hunk 2]
```diff
@@ -1161,20 +1161,25 @@ def _split_state_dict(self, state_dict):
...
             else:
-                # Split state_dict if needed
+                # Split state_dict by chunk identity
                 prefix = "model" if "model0" in state_dict.keys() else "model_"
-                offset = 0
+                chunk_id_to_global_idx = {id(chunk): idx for idx, chunk in enumerate(self.model_chunks)}
```
Comment (kunlunl) [inline megatron/core/optimizer/optimizer.py]: Why call `id(xxx)` here?

File: megatron/core/optimizer/optimizer.py (hunk 1 above)
Comment (eternally-z) [reply megatron/core/optimizer/optimizer.py:1166]: The state_dict is only read and not modified in the subsequent functions. Therefore, a deep copy is not necessary here

File: megatron/core/optimizer/optimizer.py (hunk 2 above)
Comment (eternally-z) [reply megatron/core/optimizer/optimizer.py]: I overcomplicated it by mapping by identity, overlooking that model_chunk is hashable by default. I have already removed the id() call.

**Comment by kunlunl** [comment]
@eternally-z The UT failed. Please take a look.
